subsys/mgmt/mcumgr/grp/settings_mgmt/src/settings_mgmt.c

```diff
@@ -0,0 +1,535 @@
+	key_name = (char *)malloc(key.len + 1);
+
+	if (key_name == NULL) {
+		return MGMT_ERR_ENOMEM;
+	}
+#endif
+
+	memcpy(key_name, key.value, key.len);
+	key_name[key.len] = 0;
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+	/* Send request to application to check if access should be allowed or not */
+	status = mgmt_callback_notify(MGMT_EVT_OP_SETTINGS_MGMT_ACCESS, &settings_access_data,
+				      sizeof(settings_access_data), &ret_rc, &ret_group);
+
+	if (status != MGMT_CB_OK) {
+		bool ok;
+
+		if (status == MGMT_CB_ERROR_RC) {
+			return ret_rc;
+		}
+
+		ok = smp_add_cmd_ret(zse, ret_group, (uint16_t)ret_rc);
+		goto end;
+	}
+#endif
+
+	/* Delete requested key from settings */
+	rc = settings_delete(key_name);
+
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	free(key_name);
+#endif
+
+	if (rc < 0) {
+		if (rc == -ENOENT) {
+			rc = SETTINGS_MGMT_RET_RC_KEY_NOT_FOUND;
+		} else {
+			rc = SETTINGS_MGMT_RET_RC_UNKNOWN;
+		}
+
+		ok = smp_add_cmd_ret(zse, MGMT_GROUP_ID_SETTINGS, (uint16_t)rc);
+	}
+
+end:
+	return ok ? MGMT_ERR_EOK : MGMT_ERR_EMSGSIZE;
+}
+
+/**
+ * Command handler: settings commit
+ */
+static int settings_mgmt_commit(struct smp_streamer *ctxt)
+{
+	bool ok = true;
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+	struct settings_mgmt_access settings_access_data = {
+		.access = SETTINGS_ACCESS_COMMIT,
+	};
+
+	zcbor_state_t *zse = ctxt->writer->zs;
+	enum mgmt_cb_return status;
+	int32_t ret_rc;
+	uint16_t ret_group;
+
+	/* Send request to application to check if access should be allowed or not */
+	status = mgmt_callback_notify(MGMT_EVT_OP_SETTINGS_MGMT_ACCESS, &settings_access_data,
+				      sizeof(settings_access_data), &ret_rc, &ret_group);
+
+	if (status != MGMT_CB_OK) {
+		bool ok;
+
+		if (status == MGMT_CB_ERROR_RC) {
+			return ret_rc;
+		}
+
+		ok = smp_add_cmd_ret(zse, ret_group, (uint16_t)ret_rc);
+		goto end;
+	}
+#endif
+
+	settings_commit();
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+end:
+#endif
+	return ok ? MGMT_ERR_EOK : MGMT_ERR_EMSGSIZE;
+}
+
+/**
+ * Command handler: settings load
+ */
+static int settings_mgmt_load(struct smp_streamer *ctxt)
+{
+	bool ok = true;
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+	struct settings_mgmt_access settings_access_data = {
+		.access = SETTINGS_ACCESS_LOAD,
+	};
+
+	zcbor_state_t *zse = ctxt->writer->zs;
+	enum mgmt_cb_return status;
+	int32_t ret_rc;
+	uint16_t ret_group;
+
+	/* Send request to application to check if access should be allowed or not */
+	status = mgmt_callback_notify(MGMT_EVT_OP_SETTINGS_MGMT_ACCESS, &settings_access_data,
+				      sizeof(settings_access_data), &ret_rc, &ret_group);
+
+	if (status != MGMT_CB_OK) {
+		bool ok;
+
+		if (status == MGMT_CB_ERROR_RC) {
+			return ret_rc;
+		}
+
+		ok = smp_add_cmd_ret(zse, ret_group, (uint16_t)ret_rc);
+		goto end;
+	}
+#endif
+
+	settings_load();
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+end:
+#endif
```

Comment:
I think that if you would do `if (IS_ENABLED(...)) `in 405, the compiler would not notice that `end:` is not used, and it would not have to be ifdefed.

Reply:
Fixed